### PR TITLE
Move editor plugin buttons modals to Bootstrap

### PR DESCRIPTION
**Pull request by dgrammatiko**
This is a redone for #4664 that uses the jModalClose function introduced #3918 from @okonomiyaki3000
### What is changed?

Well this PR leaves the current layouts of the editor plugin button untouched in the root folder of Joomla (this is done for B/C) but introduces the usage of bootstrap for Isis template.
The actual rendering of the buttons:

screen shot 2015-01-09 at 4 35 03
screen shot 2015-01-09 at 4 35 14
screen shot 2015-01-09 at 4 35 26

There are two questions here:

@wilsonge Shall I copy the button layout into protostar as well?

@okonomiyaki3000 I made some changes in behavior.php regarding jModalClose (due to the way bootstrap script is working):

``` php
        // Attach modal behavior to document
        $document
            ->addScriptDeclaration('
        jQuery(function($) {
            SqueezeBox.initialize(' . $options . ');
            SqueezeBox.assign($("' . $selector . '").get(), {
                parse: "rel"
            });
        });
        if(typeof jModalClose == "function"){
            var fnCode = jModalClose.toString() ;
            fnCode = fnCode.replace(/\}$/, "SqueezeBox.close();\n}");
            window.eval(fnCode);
        }
        else {
            function jModalClose() {
                SqueezeBox.close();
            }
        }'
        );
```

Is that ok? Especially this part: `window.eval(fnCode);`

### CODE REVIEW

**Comment by okonomiyaki3000** [comment]
I had in mind a different way to override the default modal with a different type. The thing is, `JHtmlBehavior::modal`, such as it is, is all about using `SqueezeBox`. If you don't want to use `SqueezeBox`, it's best to completely override that function with one of your own by using `JHtml::register`. In other words, you don't need to touch `behavior.php`.

I guess you're trying to have it such that both kinds of modals are available. I suppose I don't really agree with that idea. If you want to use bootstrap modals (good!) then why not replace `SqeezeBox` in all cases?

**Comment by dgt41** [comment]
@okonomiyaki3000 I can understand the way you meant this code to be used, totally replace the modal, and of course is a good approach. The point in this particular case, and I guess to some other places in core, is that you might end up having the core with all modals using bootstrap but some old 3pd component will use the mootools modal. 
I think this is all about B/C and how we can make different frameworks work together (especially with the upcoming layouts) 
I am afraid we end up with two solutions here:
1. Solve it in client side with some code like this one I used here
2. Solve it in server side:
Introduce a function addScriptModalClose (which creates an array similar to _scripts)
Call it and set the code for each modal
On head.php implode the array and render the proxy function

Of course I might be very wrong with my assumptions here and there might be a way easier solution. But then again thats the reason why I raise this concern...

**Comment by dgt41** [comment]
@wilsonge Can you take a look at this one (when you have some spare time, of course)? 
I would like to hear your opinion about the `jModalClose()` override and if you can think of another  possible solution for this

**Comment by dgt41** [comment]
@okonomiyaki3000 How about this approach? I end up cannibalizing the object button...

**Comment by dgt41** [comment]
This is for 3.5!!!
In order to successfully test this one you need to apply first #5652. 
@phproberto Can you take a look at this one?
The problem here is the way we handle the scripts and the closing function of bootstrap.
This might be a slight B/C issue for people already using layouts of those buttons, but I cannot figure out a totally compatible solution. Maybe you can think of something that will do the trick ♻︎

File: layouts/joomla/editors/buttons.php [hunk 1]
```diff
@@ -11,13 +11,31 @@
...
+				if (isset($button->plugin)){
+					$named = strtolower(str_replace(' ', '', $button->plugin));
+					$mpath = '/layouts/joomla/editors-xtd/' . $named . '/' . $named . '.php';
+
+					if (is_file(JPATH_THEMES . '/' . JFactory::getApplication()->getTemplate() . '/html' . $mpath)
+						|| is_file(JPATH_ROOT . $mpath))
+					{
+						$isNewLayout = 1;
+					}
+				}
```
Comment (dgt41) [inline layouts/joomla/editors/buttons.php]: @wilsonge Will the new layouts renderer have some problems with this code above?

Comment (wilsonge) [inline layouts/joomla/editors/buttons.php]: Umm what are you trying to do? If it's create template override paths for the JLayouts then we'll be doing that in JPlugin (unsure as exact paths but using the add include path(s) https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/layout/file.php#L159-L195 functions)

Comment (dgt41) [inline layouts/joomla/editors/buttons.php]: The problem is that these buttons have some js and some code for closing the modals. Right now only few params are passed to layout and then the button gets rendered. If we are about to eliminate mootools we need to set the correct javascript (modal close is the problem) in the layout. This code is a simple fall back for 3rd party plugins to keep working

**Comment by dgt41** [comment]
Needs also #5871

**Comment by nueckman** [comment]
Patch breaks ability to insert image.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/5652">issues.joomla.org/joomla-cms/5652</a>.</sub>

**Comment by dgt41** [comment]
@nueckman try to apply also #5871

**Comment by dgt41** [comment]
@nueckman works here, see the [video](https://dl.dropboxusercontent.com/u/55769984/ScreenFlow.mp4)

**Comment by dgt41** [comment]
@phproberto if you have some spare time can you review this one? Reminder it needs #5871 as well (and if you also apply #5654 you 'll drop mootools completely in from end)

**Pull request closed without merging**